### PR TITLE
fix Issue 16195 - delete should be @system

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: test/fail_compilation/test16195.d [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/*
+ * TEST_OUTPUT:
+---
+fail_compilation/test16195.d(13): Error: delete p is unsafe
+fail_compilation/test16195.d(13): Error: delete p is impure
+fail_compilation/test16195.d(13): Error: delete p is non-@nogc
```
Comment (mathias-lang-sociomantic) [inline test/fail_compilation/test16195.d]: Having a different error message here is quite confusing.

I suggest something along the lines of:
```
fail_compilation/test16195.d(13): Error: pure function 'test' cannot use impure 'delete' operator
fail_compilation/test16195.d(13): Error: safe function 'test' cannot use system 'delete' operator
fail_compilation/test16195.d(13): Error: @nogc function 'test' cannot use non-@nogc 'delete' operator
```

**Comment by WalterBright** [comment]
Blocked by https://github.com/dlang/phobos/pull/4494

File: test/fail_compilation/test16195.d [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/*
+ * TEST_OUTPUT:
+---
+fail_compilation/test16195.d(13): Error: delete p is not @safe but is used in @safe function test
+fail_compilation/test16195.d(13): Error: delete p is not @nogc but is used in @nogc function test
+fail_compilation/test16195.d(13): Error: delete p is non-@nogc
```
Comment (mathias-lang-sociomantic) [inline test/fail_compilation/test16195.d]: Looks like the test for `pure` disappeared ?

Comment (WalterBright) [inline test/fail_compilation/test16195.d]: Yup. For the same reason that `new` is not a purity violation.

File: src/expression.d [hunk 3]
```diff
@@ -11069,6 +11069,18 @@ extern (C++) final class DeleteExp : UnaExp
                 return new ErrorExp();
         }
 
+        // unsafe
+        if (!sc.intypeof && sc.func)
```
Comment (andralex) [inline src/expression.d]:
```d
if (!sc.intypeof && sc.func && sc.func.setUnsafe())
{
    error(......);
    err = true;
}
```

File: src/expression.d [hunk 4]
```diff
@@ -11069,6 +11069,18 @@ extern (C++) final class DeleteExp : UnaExp
...
+        {
+            if (sc.func.setUnsafe())
+            {
+                error("%s is not @safe but is used in @safe function %s", toChars(), sc.func.toChars());
+                err = true;
+            }
+        }
+        if (err)
```
Comment (andralex) [inline src/expression.d]: I'd use a variable `result` initiated with `this`, then assign it from `new ErrorExp` in case of error. No need for a Boolean.

**Comment by andralex** [comment]
lgtm % nits

File: src/expression.d (hunk 3 above)
Comment (WalterBright) [inline src/expression.d]: done

File: src/expression.d (hunk 4 above)
Comment (WalterBright) [inline src/expression.d]: I don't think that works out too well with the `err |= ...` used earlier. Also, frequent `new ErrorExp` makes for bloated code.

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**